### DISCUSSION TITLE
Composite image into canvas size.

### DISCUSSION THREAD

**Discussion by TNNSLAB** [ImageMagick/ImageMagick#5305] [Help] [answered]
I'm trying to understand how I can :

- read image size
- scale down if image too big
- scale up if too small
- fit into canvas

Background : 
![background](https://user-images.githubusercontent.com/46636486/177760312-86c3f008-a8f3-4cd8-95b7-d691627cf59c.png)

Desired output : 
![output](https://user-images.githubusercontent.com/46636486/177760517-06a01a37-d1e7-486b-8e28-d447691decb7.png)

Uploaded image : (too big needs ti be sized down) 
![3](https://user-images.githubusercontent.com/46636486/177761282-cfed2e51-5520-42ba-a793-f1643a3b075c.png)

Uploaded image : (too small, needs to be upscaled)
![2](https://user-images.githubusercontent.com/46636486/177761573-5ae91bd4-1cca-4a74-857d-8d0aa95f05a0.jpeg)

All I can figure out was how to composite the files.

`convert background.png uploaded_file.png -gravity northwest -geometry +312+66 -composite output.png`

**Comment by snibgo** [comment, marked as answer]
What version of IM, on what platform? Please always state this. I assume IM v7 on Windows.

Have you tried "-resize"?

Resize to *within* the required box, retaining the aspect ratio, leaving a gap at a side:
```
magick ^
  compBack.png ^
  ( compFore1.png -resize 657x334 ) ^
  -gravity northwest -geometry +310+77 -composite ^
  out1.jpg
```
![out1](https://user-images.githubusercontent.com/33812368/177811206-a0e064ce-c66c-457e-8231-51fdbc300bed.jpg)

Resize to exactly the required box, changing the aspect ratio:
```
magick ^
  compBack.png ^
  ( compFore1.png -resize "657x334^!" ) ^
  -gravity northwest -geometry +310+77 -composite ^
  out2.jpg
```
![out2](https://user-images.githubusercontent.com/33812368/177811239-6b3b118f-6ce3-479c-9496-7ec13933010e.jpg)

Crop to the aspect ratio of the required box and resize exactly, almost exactly retaining the aspect ratio:
```
magick ^
  compBack.png ^
  ( compFore1.png -gravity Center -crop 657:334+0+0 +repage -resize "657x334^!" ) ^
  -gravity northwest -geometry +310+77 -composite ^
  out3.jpg
```
![out3](https://user-images.githubusercontent.com/33812368/177811288-13ff82dd-5e64-46f7-82e1-2077b47a70b7.jpg)

**Reply by TNNSLAB** [reply]
This is exactly what I was looking for ! And again very detailed answer/s ! Fantastic ! Thank you ! 
Sorry about forgetting to mention which platform and version I use.
